wordpress-mobile/WordPress-Android
# Extracting mockable classes to add unit testing for Tenor

This is part of the #12642 and does the following:

- Extracts classes and methods so to make them mockable (4bfb9f095fe51cc991a1747b239a01d83ceca3b8) and makes the TenorGifClient injectable (8840f96d3c3985b330d415d64f08a84ac101ce4a)
- Refactors the GifMediaDataSource.kt and GifMediaInsertUseCase.kt to use the new wrapper and methods

As part of this PR I addressed this [comment](https://github.com/wordpress-mobile/WordPress-Android/pull/13104#discussion_r503861384) from #13104 . Will address [this other feedback](https://github.com/wordpress-mobile/WordPress-Android/pull/13104#discussion_r503851806) in a dedicated PR.

**NOTE:** in extracting the `TenorGifClient.kt` in the `search` function I removed the inline and crossinline keywords since it was complicating the mocking/unit testing and at least in my tests in my Pixel 3 emulator API 29 (that is pretty slow) I didn't notice appreciable differences in performance. Let me know wdyt 🙇 !

## To test
This PR should not change the behaviour introduced in #13104, so steps for testing are the same. I'm reporting the steps below from convenience.

For all the below enable the `CONSOLIDATED_MEDIA_PICKER` feature flag in wasabi.

### Media Browser insert
- In media browser, + menu -> Choose from Tenor
- Fill in search string(s) in the search and check the gifs are loaded (check they are consistent with those loaded with same string(s) in the Tenor current vanilla implementation)
- Scroll down the loaded GIFs and check the pagination works (check loaded GIFs and pages are consistent with those loaded with same string(s) in the Tenor current vanilla implementation)
- Select more than one gifs and check the selection progressive counter is consistent
- Try also unselecting some of the gifs to check the counter number stays progressive
- Confirm selection and check the upload dialog is presented with the string `Uploading gif media`
- Check you are brought back to media browser and the gifs are loaded correctly
- Smock test (like check you can preview the uploaded GIFs and similar)

### GB Image/Gallery block
- In GB add Image/Gallery block -> Choose from Tenor
- Insert a search string
- Select GIF and accept
- Note the GIF is uploaded and added to the block
- Smock test

PR submission checklist:
